File: app/api/strava/disconnect/route.ts
import { NextRequest, NextResponse } from 'next/server';
import { createServerClient } from '@supabase/ssr';
import { cookies } from 'next/headers';

export async function POST(request: NextRequest) {
  const cookieStore = await cookies();
  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        get(name: string) {
          return cookieStore.get(name)?.value;
        },
        set(name: string, value: string, options: Record<string, unknown>) {
          cookieStore.set({ name, value, ...options });
        },
        remove(name: string, options: Record<string, unknown>) {
          cookieStore.set({ name, value: '', ...options });
        },
      },
    }
  );

  // Get current user
  const { data: { user }, error: userError } = await supabase.auth.getUser();

  if (userError || !user) {
    return NextResponse.json({ error: 'Unauthorized' }, { status: 401 });
  }

  try {
    // Get the connection first to revoke the token
    const { data: connection } = await supabase
      .from('strava_connections')
      .select('access_token')
      .eq('user_id', user.id)
      .single();

    // Revoke the Strava token
    if (connection?.access_token) {
      try {
        await fetch('https://www.strava.com/oauth/deauthorize', {
          method: 'POST',
          headers: {
            'Authorization': `Bearer ${connection.access_token}`,
            'Content-Type': 'application/json',
          },
        });
      } catch (revokeError) {
        console.error('Failed to revoke Strava token:', revokeError);
        // Continue with local cleanup even if revoke fails
      }
    }

    // Delete the connection from our database
    const { error: deleteError } = await supabase
      .from('strava_connections')
      .delete()
      .eq('user_id', user.id);

    if (deleteError) {
      console.error('Failed to delete Strava connection:', deleteError);
      return NextResponse.json({
        error: 'Failed to disconnect from Strava'
      }, { status: 500 });
    }

    // Optionally, mark Strava activities as disconnected rather than deleting them
    // This preserves user's historical data
    await supabase
      .from('activities')
      .update({
        notes: 'Imported from Strava (connection now removed)',
        updated_at: new Date()
      })
      .eq('user_id', user.id)
      .eq('source', 'strava');

    return NextResponse.json({
      success: true,
      message: 'Successfully disconnected from Strava'
    });
  } catch (error) {
    console.error('Strava disconnect error:', error);
    return NextResponse.json({
      error: 'Failed to disconnect from Strava'
    }, { status: 500 });
  }
}

File: app/api/garmin/disconnect/route.py
import os
import json
from http.server import BaseHTTPRequestHandler
from supabase import create_client, Client

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        # Initialize Supabase client
        supabase_url = os.environ.get("NEXT_PUBLIC_SUPABASE_URL")
        supabase_key = os.environ.get("NEXT_PUBLIC_SUPABASE_ANON_KEY")
        supabase: Client = create_client(supabase_url, supabase_key)

        # Get user from JWT
        auth_header = self.headers.get('Authorization')
        if not auth_header or not auth_header.startswith('Bearer '):
            self.send_response(401)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'error': 'Unauthorized'}).encode())
            return

        jwt = auth_header.split(' ')[1]
        try:
            user_response = supabase.auth.get_user(jwt)
            user = user_response.user
            if not user:
                raise Exception("User not found")
        except Exception as e:
            self.send_response(401)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'error': 'Invalid token', 'details': str(e)}).encode())
            return

        # Delete connection from database
        try:
            data, count = supabase.from_('garmin_connections').delete().eq('user_id', user.id).execute()

            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'message': 'Garmin account disconnected successfully'}).encode())
        except Exception as e:
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'error': 'Database error', 'details': str(e)}).encode())
            return
        return
